### PR TITLE
Adds support for Github Branch Source plugin

### PR DESCRIPTION
**Pull request by msheiny**
This was really hacky and was lovingly lifted from an example XML groovy
conversion script I found in Github (left a comment reference in the template). Its
not really idempotent right now but it does the job of allowing us to record the
important details in code and works without intervention on first launch. Subsequent
modifications will require manual intervention (deleting existing job) but other than that
its good to go. 

Ummm how do you really test this? Well you'll have to accurately fill out two default
vars:

* `jenkins_deploy_gh_folder` AND
* `jenkins_deploy_userpass_creds`

Of particular note the `jenkins_deploy_userpass_creds`, you can call the id whatever you
want it just has to match up with the same id name you drop under `scan_cred_id` field in
the `jenkins_deploy_gh_folder` dict. You also dont need a username for the cred dict, just
the id, password, and recommended to always put a description (not sure if that last one is
really required but I always put it in).

### CODE REVIEW

File: templates/github_folder.groovy [hunk 1]
```diff
@@ -0,0 +1,108 @@
+// Adapted from https://github.com/beedemo/pse-cje-sa/quickstart/init_22_github_org_project.groovy
```
Comment (conorsch) [inline templates/github_folder.groovy]: This URL didn't work for me; https://github.com/beedemo/pse-cje-sa/blob/5e5f07529d074153440c9ff0905dd0b0889d9bfa/quickstart/init_22_github_org_project.groovy does, however.

File: tasks/config.yml [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
-  with_fileglob:
-    - templates/*groovy
+  when: item != 'skip'
+  register: result_groovy
+  with_items:
+    - "{{ 'github_folder' if jenkins_deploy_gh_folder else 'skip' }}"
+    - general_config
+    - creds
+    - "{{ 'ec2_cloud' if jenkins_deploy_ec2_clouds else 'skip' }}"
+    - "{{ 'github_auth' if jenkins_deploy_gh_oauth else 'skip' }}"
```
Comment (conorsch) [inline tasks/config.yml]: This looks like enough rope to hang yourself with, but I see the value in providing overrides.

**Comment by msheiny** [comment]
Hey @conorsch to give you some context about the bigger picture, the way a lot of these configuration pieces are working is dropping groovy init scripts that get executed on every single time jenkins boots up. I tried to avoid dropping scripts altogether if the variables werent there to set them up, or having enough jinja conditional logic in the groovy template itself.

Okay so to recap this change, if you fill out `jenkins_deploy_userpass_creds` that drops a credential in jenkins credential store. In this PR's case, we need a credential that can scan a github organization. Specifically an oauth token. That would look something like this in ansible:

```yaml
jenkins_deploy_userpass_creds:
  - id: my_custom_githuborg_id
    password: "{{ oauth_secret }}"
    description: "Good description for my credential"
```
Jenkins credential store can be accessed via the UI, and many jobs will reference this store and inject credentials in jobs when necessary.
selection_021

The bigger piece is the [github-branch-source plugin](https://wiki.jenkins-ci.org/display/JENKINS/GitHub+Branch+Source+Plugin) configuration. This plugin allows you to create dynamic job folders that will scan the organization of a github group you define, look for any repo (you can filter specific repos) that contains a `Jenkinsfile` in the root, and then proceed to run jobs against those branches. It will also dynamically register webhooks for future pushes/PRs if the credentials you provided have rights to do that. So basically, to configure that would be something like this:

```yaml
jenkins_deploy_gh_folder:
  - job_name: "GitHub Scanner"
    owner: freedomofpress
    scan_cred_id: my_custom_githuborg_id
    health_recursive: false
    # leave below blank to scan everything or specific a pattern to look for in repo names
    scan_pattern: ansible-repo 
    origin_branch: true
    origin_branch_pr: true
    origin_pr_merge: false
    origin_pr_head: false
    fork_pr_merge: false
    fork_pr_head: false
    prune_dead_branch: true
    prune_days: 0
    prune_number: 0
```

Why was this needed? Because currently to implement github hooks and scanning logic we were doing this ad-hoc via the UI and this was not saved in configuration management. This PR allows that logic to get saved in config management for easy redeployment if/when the ship blows up. You might be asking, well why dont we just take that templated XML in the middle of the groovy script and drop it into some kind of job folder? Well, jenkins gets weird when you do that sometimes and doing it through groovy is much more reliable. Its still pretty hacky though, we should be using straight groovy instead of this XML conversion logic but this was the easiest thing to go from point A -> B so we can always regroup, learn the java libraries better and write in full groovy although i dont really see the win. PHEW - sorry, does that help clear things up a bit?

**Comment by conorsch** [comment]
> You might be asking, well why dont we just take that templated XML in the middle of the groovy script and drop it into some kind of job folder? Well, jenkins gets weird when you do that sometimes and doing it through groovy is much more reliable.

Yes, that was a major question on my mind, but you managed expectations well there by forecasting that groovy would be superior to XML given Jenkins's idiosyncrasies.

> does that help clear things up a bit?

Fantastic write-up, thanks for the blow-by-blow! The org scanning is a powerful pragma: I'd expected we'd need to opt individual repos in each and every time, which would get tedious. This is way cooler.

**Pull request merged**